src/runtime/d3d12compute.cpp

```diff
@@ -2617,13 +2640,45 @@ WEAK int halide_d3d12compute_device_malloc(void *user_context, halide_buffer_t *
 WEAK int halide_d3d12compute_device_free(void *user_context, halide_buffer_t *buf) {
     TRACELOG;
 
-    TRACEPRINT("buf " << buf << " device is " << buf->device << "\n");
+    TRACEPRINT("user_context: " << user_context << " | halide_buffer_t: " << buf << " | d3d12_device: " << buf->device << "\n");
 
     if (buf->device == 0) {
         return 0;
     }
 
-    halide_d3d12compute_detach_buffer(user_context, buf);
+    d3d12_buffer *dbuffer = peel_buffer(buf);
+    TRACEPRINT("d3d12_buffer: " << dbuffer << "\n");
+
+    //wait_until_signaled(dbuffer->signal);
+
+    bool cached = false;
+    if (halide_can_reuse_device_allocations(user_context)) {
+        TRACE_SCOPE("allocation cache");
+        ScopedMutexLock lock(&buffer_pool_lock);
+        for (size_t i = 0; i < MaxBuffersInCache; ++i) {
+            if (buffer_pool[i] != NULL) {
```

Comment:
Unless I'm missing something, this cache doesn't seem to ever evict anything, so if it fills with small buffers from one part of the pipeline, you won't get any caching at all when you run a second part of the pipeline or another pipeline with larger buffers.

Is it possible to extract the cache logic from cuda.cpp and share that among the backends instead? As it stands we're going to have different logic per GPU backend for allocation caching.

Reply:
> Unless I'm missing something, this cache doesn't seem to ever evict anything, so if it fills with small buffers from one part of the pipeline, you won't get any caching at all when you run a second part of the pipeline or another pipeline with larger buffers.
> 
> Is it possible to extract the cache logic from cuda.cpp and share that among the backends instead? As it stands we're going to have different logic per GPU backend for allocation caching.

Indeed, it's a rudimentary implementation of an allocation cache as there's no LRU/aging/tracking/eviction in place. I made this mostly as a placeholder that just so happened to work relatively well for the cases we have at the moment.

I think we could adopt the one you have in CUDA, but the best thing to do (in d3d12 at least) would be to perform sub-allocations on larger ID3D12Resource heaps, instead of caching the internal buffer wrapping object (d3d12_buffer).

Reply:
Given that there's no eviction here, I think we should, for now, disable the allocation cache and re-enable once we have that.